kotlinx-coroutines-test/src/DelayController.kt

```diff
@@ -112,9 +113,75 @@ public interface DelayController {
      * Resumed dispatchers will automatically progress through all coroutines scheduled at the current time. To advance
      * time and execute coroutines scheduled in the future use, one of [advanceTimeBy],
      * or [advanceUntilIdle].
+     *
+     * When the dispatcher is resumed, all execution be immediate in the thread that triggered it. This means
+     * that the following code will not switch back from Dispatchers.IO after `withContext`
```

Comment:
While we are here, could you please elaborate on why this behaviour is preferred over "classic" dispatching?
I am a bit concerned about how this behaviour is different from "release" builds

Reply:
So the main goal of this eager behavior was to make this test function testable without having to call runCurrent():

```
fun launchesCoroutine() {
    scope.launch {
        // do stuff
    }
}
```

In a test that doesn't use multiple threads (TestCoroutineDispatcher is used for all coroutines) it provides eager entry into the launch body (during dispatch the launch body is immediately executed).

As this is both the normal structure for this code (at least on Android) it is nice to avoid extra calls to `runCurrent()` in this test

```
@Test fun callLaunchesCoroutine() = runBlockingTest {
    // assume `scope` is using this TestCoroutineDispatcher
    launchesCoroutine()
    // coroutine has already launched here
}
```

This thread switching behavior is an undesired side effect of that API choice.

So basically, I think there are two options here:

1. Modify the behavior of `dispatch` to require a call to `runCurrent()` in this test. This does make common test cases require an extra call, but leads to correct threading behavior if the developer hasn't injected `TestCoroutineDispatcher` throughout their code under test.
2. Keep the behavior (prefer eager entry of launch blocks) with the assumption most tests will not execute coroutines on multiple threads.

WDYT?

Reply:
Thanks for the explanation!

"Keep the behavior" part is, of course, more preferable as it really simplifies writing simple unit tests for end users. I've just realized that this is exactly why we thought to make `Uncofined` the default behaviour of `TestMainDispatcher`.

I think it's easier to change the docs and explain that test dispatcher acts like unconfined one (and maybe implementation should use it as well)

Reply:
That's a pretty good way to explain it. I'll add that to the docs in this PR since it's become all about threading.